File: src/main/java/ru/serdtsev/homemoney/account/model/Category.java
package ru.serdtsev.homemoney.account.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import ru.serdtsev.homemoney.account.CategoryRepository;
import ru.serdtsev.homemoney.balancesheet.BalanceSheet;

import javax.annotation.Nonnull;
import javax.persistence.*;
import java.sql.Date;
import java.util.UUID;

@Entity
@Table(name = "categories")
@DiscriminatorValue("category")
public class Category extends Account implements Comparable {
  @ManyToOne
  @JoinColumn(name = "root_id")
  private Category root;

  @Transient
  private UUID rootId;

  protected Category() {
  }

  public Category(BalanceSheet balanceSheet, AccountType type, String name, Date created, Boolean isArc, Category root) {
    super(balanceSheet, type, name, created, isArc);
    this.root = root;
  }

  public void init(CategoryRepository categoryRepo) {
    super.init();
    if (rootId != null) {
      root = categoryRepo.findOne(rootId);
    } else {
      root = null;
    }
  }

  @JsonIgnore
  public Category getRoot() {
    return root;
  }

  public UUID getRootId() {
    return rootId;
  }

  @Override
  public String getSortIndex() {
    return (root != null)
        ? root.getSortIndex() + "#" + getName()
        : Integer.toString(getType().ordinal()) + "#" + getName();
  }

  @Override
  public int compareTo(@Nonnull Object o) {
    if (!(o instanceof Category)) {
      return 0;
    }
    Category other = (Category) o;
    int typeComparing = getType().compareTo(other.getType());
    if (typeComparing != 0) {
      return typeComparing;
    }
    return getSortIndex().compareTo(other.getSortIndex());
  }
}

File: src/main/java/ru/serdtsev/homemoney/account/model/Category.kt
package ru.serdtsev.homemoney.account.model

import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.data.repository.findByIdOrNull
import ru.serdtsev.homemoney.account.CategoryRepository
import ru.serdtsev.homemoney.balancesheet.BalanceSheet
import java.sql.Date
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "categories")
@DiscriminatorValue("category")
open class Category(
        id: UUID,
        balanceSheet: BalanceSheet,
        type: AccountType,
        name: String,
        created: Date,
        isArc: Boolean? = null,
        @JsonIgnore @ManyToOne @JoinColumn(name = "root_id") open var root: Category? = null
) : Account(id, balanceSheet, type, name, created, isArc), Comparable<Account> {

    @Transient
    open val rootId: UUID? = null

    fun init(categoryRepo: CategoryRepository) {
        root = rootId?.let { categoryRepo.findByIdOrNull(rootId)!! }
    }

    override fun getSortIndex(): String {
        return root?.let { "${it.getSortIndex()}#$name" } ?: "${type.ordinal}#$name"
    }

    override fun compareTo(other: Account): Int {
        if (other !is Category) {
            return 0
        }
        val typeComparing = type.compareTo(other.type)
        return if (typeComparing != 0) typeComparing else getSortIndex().compareTo(other.getSortIndex())
    }
}